source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -420,34 +410,21 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, object> allocateI
             sb.AppendLine(statement);
             sb.AppendLine();
 
-            if (relatedDocumentData.Any(d => d.Related.Any()))
-                sb.AppendLine("DECLARE @references as TABLE (Reference nvarchar(400), ReferenceTable nvarchar(400))");
-
+            var index = 0;
             foreach (var data in relatedDocumentData)
             {
                 if (data.Related.Any())
                 {
-                    var relatedVariablePrefix = $"{data.TableName.ToLower()}_";
-
-                    sb.AppendLine();
-                    sb.AppendLine("DELETE FROM @references");
-                    sb.AppendLine();
-
-                    var valueBlocks = data.Related.Select((r, idx) => $"(@{relatedVariablePrefix}{idx}, '{r.relatedTableName}')");
-                    sb.Append("INSERT INTO @references VALUES ");
-                    sb.AppendLine(string.Join(", ", valueBlocks));
-                    sb.AppendLine();
+                    var tableVariableName = $"{index++}__relatedDocumentTableValuedParameter";
+                    parameters.AddTable(tableVariableName, CreateRelatedDocumentTableValuedParameter(mapping, data));
 
                     sb.AppendLine($"DELETE FROM [{data.SchemaName}].[{data.TableName}] WHERE [{data.IdColumnName}] = @{IdVariableName}");
-                    sb.AppendLine($"    AND [{data.RelatedDocumentIdColumnName}] not in (SELECT Reference FROM @references)");
+                    sb.AppendLine($"    AND [{data.RelatedDocumentIdColumnName}] not in (SELECT [{data.RelatedDocumentIdColumnName}] FROM @{tableVariableName})");
```

Comment:
We can use our table valued parameter right here, instead of populating a temp table 😂